product_code/certificate_service/impl/src/main/scala/de/upb/cs/uc4/certificate/impl/CertificateServiceImpl.scala

```diff
@@ -0,0 +1,124 @@
+import de.upb.cs.uc4.certificate.model.{ EncryptedPrivateKey, JsonCertificate, JsonEnrollmentId, PostMessageCSR }
+import de.upb.cs.uc4.hyperledger.HyperledgerAdminParts
+import de.upb.cs.uc4.hyperledger.utilities.traits.EnrollmentManagerTrait
+import de.upb.cs.uc4.shared.client.exceptions.{ DetailedError, ErrorType, UC4Exception }
+import de.upb.cs.uc4.shared.server.ServiceCallFactory._
+import de.upb.cs.uc4.shared.server.messages.{ Accepted, Confirmation }
+
+import scala.concurrent.duration._
+import scala.concurrent.{ Await, ExecutionContext, Future, TimeoutException }
+
+/** Implementation of the UserService */
+class CertificateServiceImpl(
+    clusterSharding: ClusterSharding,
+    enrollmentManager: EnrollmentManagerTrait,
+    readSide: ReadSide, processor: CertificateEventProcessor
+)(implicit ec: ExecutionContext, val config: Config)
+  extends CertificateService with HyperledgerAdminParts {
+  readSide.register(processor)
+
+  /** Looks up the entity for the given ID */
+  private def entityRef(id: String): EntityRef[CertificateCommand] =
+    clusterSharding.entityRefFor(CertificateState.typeKey, id)
+
+  implicit val timeout: Timeout = Timeout(15.seconds)
+
+  lazy val validationTimeout: FiniteDuration = config.getInt("uc4.validation.timeout").milliseconds
+
+  /** Forwards the certificate signing request from the given user */
+  override def setCertificate(username: String): ServerServiceCall[PostMessageCSR, Done] = identifiedAuthenticated(AuthenticationRole.All: _*) {
+    (authUsername, _) =>
+      ServerServiceCall { (_, pmcsrRaw) =>
+        // One can only set his own certificate
+        if (authUsername != username) {
+          throw UC4Exception.OwnerMismatch
+        }
+
+        // Validate the CSR
+        val validationErrorsFuture = pmcsrRaw.validate
+        val validationErrors = try {
+          Await.result(validationErrorsFuture, validationTimeout)
+        }
+        catch {
+          case _: TimeoutException => throw UC4Exception.ValidationTimeout
+          case e: Exception        => throw UC4Exception.InternalServerError("Validation Error", e.getMessage)
+        }
+        if (validationErrors.nonEmpty) {
+          throw new UC4Exception(422, DetailedError(ErrorType.Validation, validationErrors))
+        }
+
+        getCertificateUser(username).flatMap {
+          case (Some(enrollmentId), Some(enrollmentSecret), None, None) =>
+            val certificate = enrollmentManager.enrollSecure(caURL, tlsCert, enrollmentId, enrollmentSecret, pmcsrRaw.certificateSigningRequest, adminUsername, walletPath, channel, chaincode, networkDescriptionPath)
+            entityRef(username).ask[Confirmation](replyTo => SetCertificateAndKey(certificate, pmcsrRaw.encryptedPrivateKey, replyTo)).map {
+              case Accepted     => (ResponseHeader(202, MessageProtocol.empty, List()), Done)
+              case e: Exception => throw UC4Exception.InternalServerError("Failed to set certificate", e.getMessage)
```

Comment:
This is false.
You will never get an exception here.
You will get a RejectedWithReason.

Reply:
Good point, this is now resolved